Add PackSegmentsFix operator in caffe2

nateanl opened the pull request:
Summary:
This operator is similar to PackSegmentsOp but with an additional option dim.
Input:  a UNKNOWN_DIM * D tensor and a lengths tensor. The shape of lengths is UNKNOWN_DIM
The output will be UNKNOWN_DIM  * dim * D

Differential Revision: D8702782

bddppq commented:
IIUC what you wanna achieve is to use an argument (aka. current named `dim`, I suggest to rename it to `max_length`) to override the max_length dimension in the output? If so you can simply add this argument to the existing PackSegmentsOp operator.

bddppq reviewed (commented):
IIUC what you wanna achieve is to use an argument (currently named dim, I suggest to rename it to max_length) to override the max_length dimension in the output? If so you can simply add this argument to the existing PackSegmentsOp operator.

bddppq commented on caffe2/operators/pack_segments.cc:
```diff
@@ -32,6 +32,11 @@ bool PackSegmentsOp<CPUContext>::DoRunWithType2() {
   }
+  if (max_length_ != -1) {
+    // Final dim must be greater than the max_length
+    CAFFE_ENFORCE_GE(max_length_, max_length, "DATA should be at least 1-D");
```
error message is inaccurate

nateanl replied:
You are right.

bddppq commented on caffe2/operators/pack_segments.h:
```diff
@@ -45,6 +46,7 @@ class PackSegmentsOp final : public Operator<Context> {
   INPUT_TAGS(LENGTHS, DATA);
 
  private:
+  int max_length_;
```
use TIndex (aka int64_t)

bddppq commented on caffe2/operators/pack_segments.cu:
```diff
@@ -186,15 +186,31 @@ bool PackSegmentsOp<CUDAContext>::DoRunWithType2() {
-                                         context_)
-                                   : 0;
 
+  T temp;
```
why here needs to be changed to call array_max  in both branches of an if-else? can't it be changed like in cpu implementation?
```
if (max_length_ != -1) {
    // Final dim must be greater than the max_length
    CAFFE_ENFORCE_GE(max_length_, max_length, "...");
    max_length = max_length_;
}
```

nateanl replied:
I guess I wanted to deal with the const that time. I will change it

bddppq commented on caffe2/python/operator_test/pack_ops_test.py:
```diff
@@ -49,6 +49,93 @@ def pack_segments_ref(lengths, data):
 
         return pack_segments_ref
 
+    def pack_segments_with_max_length_ref(self, return_presence_mask=False):
```
could you extend `pack_segments_ref` to take an additional optional max_length argument?

nateanl replied:
You mean delete this function and modify the existing one?

bddppq replied:
yep

bddppq commented on caffe2/python/operator_test/pack_ops_test.py:
```diff
@@ -14,15 +14,19 @@
 
-    def pack_segments_ref(self, return_presence_mask=False):
-        def pack_segments_ref(lengths, data):
+    def pack_segments_ref(self, return_presence_mask=False, max_length=-1):
```
nit: maybe max_length=None? In Python it's common to use None as default value to indicate optional argument.

bddppq commented:
@pytorchbot retest this please

bddppq reviewed (approved):
LGTM

bddppq commented on caffe2/python/operator_test/pack_ops_test.py:
```diff
@@ -49,6 +53,58 @@ def pack_segments_ref(lengths, data):
+            + "=" * 60
+        )
+        # run test
+        max_length = 100
```
`num_seq` can be 100 (since `st.integers` can return value <= max_value), maybe set it to `num_seq + 1`?

nateanl replied:
Yes, I agree.

nateanl committed:
Modify the original PackSegments operator by adding "max_length" argument (#9048)

Summary:
Closes https://github.com/pytorch/pytorch/pull/9048

max_length argument helps fix the shape of the output to be N * max_length * D, where N is the batch_size, D is the feature_dim.

Differential Revision: D8702782

fbshipit-source-id: 599dd40821a5ca119acb4d61162f194f222cfbfb
